pubsub/natspubsub/nats_streaming.go

```diff
@@ -0,0 +1,277 @@
+package natspubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/stan.go"
+
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+type stmTopic struct {
+	nc   stan.Conn
+	subj string
+}
+
+type StreamingTopicOptions struct {
+}
+
+// OpenStreamingTopic returns a *pubsub.Topic for use with NATS Streaming Server.
+// The subject is the NATS Subject; for more info, see
+// https://nats.io/documentation/writing_applications/subjects.
+func OpenStreamingTopic(nc stan.Conn, subject string, _ *StreamingTopicOptions) (*pubsub.Topic, error) {
+	dt, err := openStreamingTopic(nc, subject)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(dt, nil), nil
+}
+
+func openStreamingTopic(nc stan.Conn, subject string) (driver.Topic, error) {
+	if nc == nil {
+		return nil, errors.New("natspubsub: stan.Conn is required")
+	}
+
+	return &stmTopic{nc, subject}, nil
+}
+
+func (t *stmTopic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
+	if t == nil || t.nc == nil {
+		return errNotInitialized
+	}
+
+	for _, m := range msgs {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		payload, err := encodeMessage(m)
+		if err != nil {
+			return err
+		}
+
+		if m.BeforeSend != nil {
+			asFunc := func(i interface{}) bool { return false }
+			if err := m.BeforeSend(asFunc); err != nil {
+				return err
+			}
+		}
+
+		if err := t.nc.Publish(t.subj, payload); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *stmTopic) IsRetryable(err error) bool {
+	return false
+}
+
+func (t *stmTopic) As(i interface{}) bool {
+	c, ok := i.(*stan.Conn)
+	if !ok {
+		return false
+	}
+	*c = t.nc
+	return true
+}
+
+func (t *stmTopic) ErrorAs(error, interface{}) bool {
+	return false
+}
+
+func (t *stmTopic) ErrorCode(err error) gcerrors.ErrorCode {
+	switch err {
+	case nil:
+		return gcerrors.OK
+	case context.Canceled:
+		return gcerrors.Canceled
+	case errNotInitialized:
+		return gcerrors.NotFound
+	case nats.ErrBadSubject, stan.ErrBadConnection:
+		return gcerrors.FailedPrecondition
+	case nats.ErrAuthorization:
+		return gcerrors.PermissionDenied
+	case nats.ErrMaxPayload, nats.ErrReconnectBufExceeded:
+		return gcerrors.ResourceExhausted
+	case stan.ErrTimeout:
+		return gcerrors.DeadlineExceeded
+	}
+	return gcerrors.Unknown
+}
+
+func (t *stmTopic) Close() error {
+	return nil
+}
+
+type stmSub struct {
+	nc       stan.Conn
+	sub      stan.Subscription
+	ack      bool
+	messages []*stan.Msg
+	sync.Mutex
+}
+
+// OpenStreamingSubscription returns a *pubsub.Subscription representing a NATS Streaming subscription.
+// The subject is the NATS Subject to subscribe to; for more info, see
+// https://nats.io/documentation/writing_applications/subjects.
+func OpenStreamingSubscription(nc stan.Conn, subject, queue string, oo ...stan.SubscriptionOption) (*pubsub.Subscription, error) {
```

Comment:
Please create a separate `Options` struct, in case we need to add options that are not `stan.SubscriptionOptions`.

Reply:
Should it contain all the options of `stan.SubscriptionOptions` as fields or `stan.SubscriptionOptions` as a field?